# Basic SQL Data Exploration

I wanted to include something that actually used SQL rather than just listing it as a bullet point on my resume. I started using a cloud instance of IBM's DB2, but figured it might be fun to explore a flavor I haven't used before. I set up a basic PostgreSQL database to do play around with. 

I found a dataset on Kaggle that included books, users, and reviews in a few separate files and thought it might be be a good candidate to play around with since we can create multiple tables. Here's a link to that dataset. 

https://www.kaggle.com/datasets/saurabhbagchi/books-dataset

I'll include the script that read those csv files in the repository.

As for what we want to do... Let's use Python and extend it's capability with sqlalchemy and sql magic to connect to our database and explore our data. Let's answer a few questions about the data that might be interesting. Let's start with the following:

1. What users have written the most reviews?
2. Which users tend to write the highest reviews? What about users writing the lowest reviews?
3. Which books are the most highly reviewed?
4. Which authors are most highly reviewed?
5. Is there any trend in rating vs published year?
6. How many books were published each year? 
7. Which publishers release the most books per year?
8. Which publishers are rated the highest and lowest?
9. Which authors have the most books?
10. Which authors have the most reviews? Is there any correlation to the authors with the most books?

I'll include all queries for these questions in a separate sql script as well.

In [181]:
# load iPython sql extension and sqlalchemy
%load_ext sql
from sqlalchemy import create_engine

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [188]:
# connect to local database.
# error shown since password was purposely obscured for upload
%sql postgresql://postgres:password123@localhost/postgres

(psycopg2.OperationalError) connection to server at "localhost" (::1), port 5432 failed: FATAL:  password authentication failed for user "postgres"

(Background on this error at: https://sqlalche.me/e/14/e3q8)
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys(['postgresql://postgres:***@localhost/postgres'])


In [3]:
# prep sqlalchemy engine
create_engine('postgresql://postgres:0829alAL@localhost/postgres')

Engine(postgresql://postgres:***@localhost/postgres)

First, let's just look at what each table looks like and understand the amount of data we are working with.

There are three tables: books, ratings, and users.

In [4]:
%%sql
select table_name 
from information_schema.tables
where table_schema = 'public';

 * postgresql://postgres:***@localhost/postgres
3 rows affected.


[('ratings',), ('books',), ('users',)]

In [5]:
%sql select * from books limit 5;

 * postgresql://postgres:***@localhost/postgres
5 rows affected.


[('0195153448', 'Classical Mythology', 'Mark P. O. Morford', 2002, 'Oxford University Press', 'http://images.amazon.com/images/P/0195153448.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0195153448.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0195153448.01.LZZZZZZZ.jpg'),
 ('0002005018', 'Clara Callan', 'Richard Bruce Wright', 2001, 'HarperFlamingo Canada', 'http://images.amazon.com/images/P/0002005018.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0002005018.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0002005018.01.LZZZZZZZ.jpg'),
 ('0060973129', 'Decision in Normandy', "Carlo D'Este", 1991, 'HarperPerennial', 'http://images.amazon.com/images/P/0060973129.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0060973129.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0060973129.01.LZZZZZZZ.jpg'),
 ('0374157065', 'Flu: The Story of the Great Influenza Pandemic of 1918 and the Search for the Virus That Caused It', 'Gina Bari Kolata', 1999, 'Farrar Straus Giroux', 'http://images.amazon.com/images/P/0374157065.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0374157065.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0374157065.01.LZZZZZZZ.jpg'),
 ('0393045218', 'The Mummies of Urumchi', 'E. J. W. Barber', 1999, 'W. W. Norton &amp; Company', 'http://images.amazon.com/images/P/0393045218.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0393045218.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0393045218.01.LZZZZZZZ.jpg')]

In [6]:
%sql select * from ratings limit 5;

 * postgresql://postgres:***@localhost/postgres
5 rows affected.


[(276725, '034545104X', 0),
 (276726, '0155061224', 5),
 (276727, '0446520802', 0),
 (276729, '052165615X', 3),
 (276729, '0521795028', 6)]

In [7]:
%sql select * from users limit 5;

 * postgresql://postgres:***@localhost/postgres
5 rows affected.


[(1, 'nyc, new york, usa', None),
 (2, 'stockton, california, usa', 18),
 (3, 'moscow, yukon territory, russia', None),
 (4, 'porto, v.n.gaia, portugal', 17),
 (5, 'farnborough, hants, united kingdom', None)]

In [12]:
%%sql 
select 
(select count(*) from books) as book_count,
(select count(*) from users) as user_count,
(select count(*) from ratings) as rating_count

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[(271379, 278858, 1149780)]

Now we can start create some basic queries to answer our questions.

### 1. What users have written the most reviews?

In [25]:
%%sql
select user_id, count(user_id)
from ratings
group by user_id
order by count(user_id) desc
limit 20;

 * postgresql://postgres:***@localhost/postgres
20 rows affected.


[(11676, 13602),
 (198711, 7550),
 (153662, 6109),
 (98391, 5891),
 (35859, 5850),
 (212898, 4785),
 (278418, 4533),
 (76352, 3367),
 (110973, 3100),
 (235105, 3067),
 (230522, 2991),
 (16795, 2948),
 (234623, 2674),
 (36836, 2529),
 (52584, 2512),
 (245963, 2507),
 (204864, 2504),
 (55492, 2459),
 (185233, 2448),
 (171118, 2421)]

Now, in any real world analysis, we would ask how someone could reasonably review so many books. I believe this is a real dataset, so it looks like we have users reviewing books they have not read.

Just for perspective, if someone read one entire book a day, it would take more than 37 years to read 13602 books... But, outliers are generally where the interesting stuff is happening. 

That's outside the scope of this exercise so let's continue.

### 2. Which users tend to write the highest reviews? What about users writing the lowest reviews?

In [49]:
%%sql
select user_id, avg(rating) as average_rating, count(user_id) as num_of_reviews
from ratings
group by user_id
order by avg(rating) desc
limit 10;

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[(14291, Decimal('10.0000000000000000'), 1),
 (140766, Decimal('10.0000000000000000'), 1),
 (68121, Decimal('10.0000000000000000'), 1),
 (136947, Decimal('10.0000000000000000'), 1),
 (106753, Decimal('10.0000000000000000'), 1),
 (128081, Decimal('10.0000000000000000'), 1),
 (30256, Decimal('10.0000000000000000'), 2),
 (17146, Decimal('10.0000000000000000'), 1),
 (71079, Decimal('10.0000000000000000'), 1),
 (136361, Decimal('10.0000000000000000'), 1)]

This is what we would expect. Folks review things they really love and nothing else. Or maybe a friend wrote the book so they left a postitive review. Let's continue but add the stipulation that the users need more than ten reviews.

In [57]:
%%sql
select user_id, round(avg(rating), 2) as average_rating, count(user_id) as num_of_reviews
from ratings
group by user_id
having count(user_id) > 10
order by avg(rating) desc
limit 10;

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[(256765, Decimal('10.00'), 16),
 (78469, Decimal('10.00'), 18),
 (86263, Decimal('10.00'), 12),
 (226362, Decimal('9.95'), 20),
 (73399, Decimal('9.90'), 21),
 (51207, Decimal('9.83'), 12),
 (102154, Decimal('9.83'), 83),
 (137222, Decimal('9.78'), 23),
 (753, Decimal('9.75'), 16),
 (37168, Decimal('9.71'), 17)]

In [64]:
%%sql
-- now for lowest ratings, we will remove folks that only leave negative reviews here
-- even though that seems that is still what is occuring here
select user_id, round(avg(rating), 2) as average_rating, count(user_id) as num_of_reviews
from ratings
group by user_id
having count(user_id) > 10 and avg(rating) > 0
order by avg(rating)
limit 10;

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[(212898, Decimal('0.00'), 4785),
 (127429, Decimal('0.01'), 1382),
 (60277, Decimal('0.01'), 614),
 (53430, Decimal('0.02'), 63),
 (198711, Decimal('0.02'), 7550),
 (82926, Decimal('0.02'), 228),
 (87746, Decimal('0.02'), 810),
 (245839, Decimal('0.03'), 230),
 (159376, Decimal('0.03'), 229),
 (242824, Decimal('0.03'), 1747)]

### 3. Which books are the most highly reviewed?

In [72]:
%%sql
select books.isbn, book_title, book_author, publisher, avg(rating), count(rating)
from books
inner join ratings on books.isbn = ratings.isbn
group by books.isbn
order by avg(rating) desc
limit 10;

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[('0001953877', 'Collins guide to dinosaurs', 'David Lambert', 'Collins', Decimal('10.0000000000000000'), 1),
 ('0002187272', 'Mr. Wright', 'Ian Wright', 'HarperCollinsWillow', Decimal('10.0000000000000000'), 1),
 ('0001935968', 'The Brambly Hedge Treasury', 'Jill Barklem', 'HarperCollins Publishers', Decimal('10.0000000000000000'), 1),
 ('0001821326', 'Paddington at the Tower (A Paddington Picture Book)', 'Michael Bond', 'Collin', Decimal('10.0000000000000000'), 1),
 ('0001857258', "Little Wolf's Book of Badness", 'Ian Whybrow', 'Collins', Decimal('10.0000000000000000'), 1),
 ('0001981323', "The Badger's Bath (Percy's Park)", 'Nick Butterworth', 'HarperCollins Publishers', Decimal('10.0000000000000000'), 1),
 ('0001360469', 'Babe Dressing', 'Mandy Stanley', 'HarperCollins Publishers', Decimal('10.0000000000000000'), 1),
 ('0001845039', 'The Moon of Gomrath', 'Alan Garner', 'HarperCollins Publishers', Decimal('10.0000000000000000'), 1),
 ('0001374869', 'Baby Plays (Collins Baby and Toddler Series)', 'Fiona Pragoff', 'HarperCollins Publishers', Decimal('10.0000000000000000'), 1),
 ('0002191792', 'Birds of East Africa (Collins Field Guides)', 'John G. Williams', 'Trafalgar Square', Decimal('10.0000000000000000'), 1)]

I added the count(rating) column since I figured we would run into a situation where the books with just a few great reviews we prevail. Let's correct this.

We can also change things up and require 50 reviews to as an acceptable threshold. Keep in mind most queries in this notebook would change based on slight alterations to this number.

In [81]:
%%sql
select books.isbn, book_title, book_author, avg(rating) as average_rating, count(rating) as num_of_ratings
from books
inner join ratings on books.isbn = ratings.isbn
group by books.isbn
having count(rating) > 50
order by avg(rating) desc
limit 10;

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[('1844262553', 'Free', 'Paul Vincent', Decimal('7.9629629629629630'), 54),
 ('0439064864', 'Harry Potter and the Chamber of Secrets (Book 2)', 'J. K. Rowling', Decimal('6.6117647058823529'), 170),
 ('0439139597', 'Harry Potter and the Goblet of Fire (Book 4)', 'J. K. Rowling', Decimal('6.5412371134020619'), 194),
 ('0439136350', 'Harry Potter and the Prisoner of Azkaban (Book 3)', 'J. K. Rowling', Decimal('6.4670050761421320'), 197),
 ('0590353403', "Harry Potter and the Sorcerer's Stone (Book 1)", 'J. K. Rowling', Decimal('6.3630952380952381'), 168),
 ('0618002227', 'The Fellowship of the Ring (The Lord of the Rings, Part 1)', 'J. R. R. Tolkien', Decimal('6.2063492063492063'), 63),
 ('0385199570', 'The Stand (The Complete and Uncut Edition)', 'Stephen King', Decimal('6.1754385964912281'), 57),
 ('0877017883', 'Griffin &amp; Sabine: An Extraordinary Correspondence', 'Nick Bantock', Decimal('6.0416666666666667'), 72),
 ('0156528207', 'The Little Prince', 'Antoine de Saint-Exupéry', Decimal('5.7974683544303797'), 79),
 ('039480001X', 'The Cat in the Hat', 'Dr. Seuss', Decimal('5.7307692307692308'), 52)]

### 4. Which authors are the most highly reviewed?

Once again, we need to specify a minimum amount of reviews to make this worthwhile.

In [86]:
%%sql
select book_author, avg(rating) as average_rating, count(rating) as num_of_ratings
from books
inner join ratings on books.isbn = ratings.isbn
group by book_author
having count(rating) > 50
order by avg(rating) desc
limit 25;

 * postgresql://postgres:***@localhost/postgres
25 rows affected.


[('Paul Vincent', Decimal('7.9454545454545455'), 55),
 ('Clamp', Decimal('6.5133333333333333'), 150),
 ('Rumiko Takahashi', Decimal('6.4457831325301205'), 83),
 ('Joanne K. Rowling', Decimal('6.2786885245901639'), 61),
 ('Bill Amend', Decimal('6.2641509433962264'), 53),
 ('Harlan Ellison', Decimal('6.1340206185567010'), 97),
 ('Shel Silverstein', Decimal('6.1058201058201058'), 189),
 ('G. B. Trudeau', Decimal('5.9322033898305085'), 59),
 ('Robert N. Munsch', Decimal('5.8823529411764706'), 51),
 ('Federico Garcia Lorca', Decimal('5.8750000000000000'), 56),
 ('Garth Ennis', Decimal('5.6142857142857143'), 70),
 ('Bill Watterson', Decimal('5.5618181818181818'), 550),
 ('Antoine de Saint-Exupéry', Decimal('5.5539568345323741'), 278),
 ('Herge', Decimal('5.5197368421052632'), 152),
 ('HUNTER S. THOMPSON', Decimal('5.4531250000000000'), 64),
 ('Jane Belk Moncure', Decimal('5.4411764705882353'), 68),
 ('Bradley Trevor Greive', Decimal('5.4316546762589928'), 139),
 ('J. K. Rowling', Decimal('5.3858321870701513'), 2908),
 ('Nick Bantock', Decimal('5.2925925925925926'), 270),
 ('Mary Engelbreit', Decimal('5.2528735632183908'), 87),
 ('Pablo Neruda', Decimal('5.2417582417582418'), 91),
 ('Alan Moore', Decimal('5.2150537634408602'), 93),
 ('Suzy Becker', Decimal('5.2028985507246377'), 69),
 ('DAN SIMMONS', Decimal('5.1898734177215190'), 79),
 ('J. R. R. Tolkien', Decimal('5.1656441717791411'), 978)]

The data definitely have a few issues. For example, Joanne K. Rowling and J.K. Rowling are the same person. In any real world scenario we would need to deal with, which can be a bit tedious so we will consider outside the scope of this project.

As a side note, some of the same books are listed multiple times with different translations, so we would need to decide how we want to interpret or clean those data as well.

Below is just a sample query to show how many different spellings for one author we would need to deal with.

In [90]:
%%sql
select distinct book_author from books where book_author like '%Rowling';

 * postgresql://postgres:***@localhost/postgres
5 rows affected.


[('J .K. Rowling',),
 ('J. K. Rowling',),
 ('J.K. Rowling',),
 ('Joanne K. Rowling',),
 ('Marjorie Rowling',)]

### 5. Is there any trend in rating vs published year?

In [111]:
%%sql
-- select the average book rating per year
select b.pub_year, avg(r.rating) as average_rating, 
       count(r.rating) as num_of_ratings
from books b
inner join ratings r on b.isbn = r.isbn
group by b.pub_year
having b.pub_year <= 2021 and b.pub_year > 0;

 * postgresql://postgres:***@localhost/postgres
109 rows affected.


[(1376, Decimal('4.0000000000000000'), 1),
 (1378, Decimal('10.0000000000000000'), 1),
 (1806, Decimal('5.0000000000000000'), 1),
 (1897, Decimal('0E-20'), 1),
 (1900, Decimal('1.7500000000000000'), 4),
 (1901, Decimal('4.5714285714285714'), 7),
 (1902, Decimal('1.8000000000000000'), 10),
 (1904, Decimal('10.0000000000000000'), 1),
 (1906, Decimal('5.0000000000000000'), 1),
 (1908, Decimal('10.0000000000000000'), 3),
 (1909, Decimal('0E-20'), 2),
 (1910, Decimal('0E-20'), 1),
 (1911, Decimal('2.1153846153846154'), 26),
 (1914, Decimal('0E-20'), 1),
 (1917, Decimal('0E-20'), 1),
 (1919, Decimal('0E-20'), 1),
 (1920, Decimal('3.1500000000000000'), 100),
 (1921, Decimal('0E-20'), 3),
 (1922, Decimal('0E-20'), 2),
 (1923, Decimal('3.1707317073170732'), 41),
 (1924, Decimal('2.6666666666666667'), 3),
 (1925, Decimal('3.6000000000000000'), 5),
 (1926, Decimal('2.3636363636363636'), 11),
 (1927, Decimal('3.0500000000000000'), 20),
 (1928, Decimal('3.2500000000000000'), 4),
 (1929, Decimal('4.0416666666666667'), 24),
 (1930, Decimal('3.9552238805970149'), 67),
 (1931, Decimal('4.0000000000000000'), 3),
 (1932, Decimal('2.0476190476190476'), 21),
 (1933, Decimal('1.2500000000000000'), 4),
 (1934, Decimal('5.0000000000000000'), 1),
 (1935, Decimal('4.3333333333333333'), 3),
 (1936, Decimal('4.4000000000000000'), 20),
 (1937, Decimal('2.6250000000000000'), 16),
 (1938, Decimal('6.3125000000000000'), 16),
 (1939, Decimal('4.3000000000000000'), 10),
 (1940, Decimal('2.4561403508771930'), 57),
 (1941, Decimal('2.2105263157894737'), 19),
 (1942, Decimal('2.7446808510638298'), 47),
 (1943, Decimal('2.7916666666666667'), 24),
 (1944, Decimal('3.7500000000000000'), 4),
 (1945, Decimal('4.0740740740740741'), 27),
 (1946, Decimal('4.0638297872340426'), 47),
 (1947, Decimal('3.2692307692307692'), 26),
 (1948, Decimal('4.9130434782608696'), 23),
 (1949, Decimal('4.7142857142857143'), 21),
 (1950, Decimal('3.3906250000000000'), 64),
 (1951, Decimal('3.3163265306122449'), 98),
 (1952, Decimal('4.5569620253164557'), 79),
 (1953, Decimal('3.6118299445471349'), 541),
 (1954, Decimal('3.1129707112970711'), 239),
 (1955, Decimal('3.4578947368421053'), 190),
 (1956, Decimal('2.9702380952380952'), 168),
 (1957, Decimal('3.8725868725868726'), 259),
 (1958, Decimal('3.3592233009708738'), 206),
 (1959, Decimal('3.4898785425101215'), 494),
 (1960, Decimal('3.4220963172804533'), 353),
 (1961, Decimal('3.4768786127167630'), 346),
 (1962, Decimal('3.2688172043010753'), 279),
 (1963, Decimal('3.5094339622641509'), 318),
 (1964, Decimal('3.1401617250673854'), 371),
 (1965, Decimal('3.0807560137457045'), 582),
 (1966, Decimal('3.1280898876404494'), 445),
 (1967, Decimal('3.0000000000000000'), 343),
 (1968, Decimal('3.4756335282651072'), 513),
 (1969, Decimal('2.9010727056019070'), 839),
 (1970, Decimal('2.7795334838224229'), 1329),
 (1971, Decimal('3.1077199281867145'), 1114),
 (1972, Decimal('2.9116607773851590'), 1698),
 (1973, Decimal('2.7626918536009445'), 1694),
 (1974, Decimal('2.8803142329020333'), 2164),
 (1975, Decimal('2.6925301204819277'), 2075),
 (1976, Decimal('2.8463521526166538'), 3879),
 (1977, Decimal('2.6716952186141652'), 3911),
 (1978, Decimal('2.6073405782243182'), 4877),
 (1979, Decimal('2.5293188994136220'), 4434),
 (1980, Decimal('2.4331473816807630'), 5557),
 (1981, Decimal('2.4868750824429495'), 7581),
 (1982, Decimal('2.3851390220517737'), 10430),
 (1983, Decimal('2.4928777274598600'), 12145),
 (1984, Decimal('2.5601187889323483'), 13806),
 (1985, Decimal('2.4367281708557315'), 13766),
 (1986, Decimal('2.5528797674832434'), 16859),
 (1987, Decimal('2.4863992379339543'), 18896),
 (1988, Decimal('2.5096471489443194'), 20887),
 (1989, Decimal('2.5199284720842440'), 25165),
 (1990, Decimal('2.6228317077960333'), 30554),
 (1991, Decimal('2.6395085874946519'), 32722),
 (1992, Decimal('2.6711939535148037'), 35194),
 (1993, Decimal('2.6468709086148206'), 38190),
 (1994, Decimal('2.7205021100680389'), 46444),
 (1995, Decimal('2.6607509289556874'), 5

I won't waste time plotting this because the answer is no, there isn't notable correlation between published year and rating.

### 6. How many books were published each year?

In [189]:
%%sql
select pub_year, count(isbn)
from books b
group by pub_year
having pub_year > 0 and pub_year <= 2021
order by count(isbn) desc;

 * postgresql://postgres:***@localhost/postgres
109 rows affected.


[(2002, 17628),
 (1999, 17432),
 (2001, 17360),
 (2000, 17235),
 (1998, 15767),
 (1997, 14892),
 (2003, 14359),
 (1996, 14031),
 (1995, 13548),
 (1994, 11796),
 (1993, 10602),
 (1992, 9906),
 (1991, 9389),
 (1990, 8661),
 (1989, 7937),
 (1988, 7493),
 (1987, 6529),
 (1986, 5841),
 (2004, 5839),
 (1985, 5343),
 (1984, 4986),
 (1983, 4499),
 (1982, 4197),
 (1981, 3279),
 (1980, 2676),
 (1979, 2208),
 (1978, 2131),
 (1977, 1897),
 (1976, 1599),
 (1975, 1219),
 (1974, 1024),
 (1973, 924),
 (1972, 773),
 (1971, 540),
 (1970, 459),
 (1969, 339),
 (1968, 233),
 (1966, 183),
 (1967, 176),
 (1965, 173),
 (1964, 149),
 (1960, 133),
 (1963, 132),
 (1961, 132),
 (1962, 123),
 (1959, 103),
 (1958, 77),
 (1957, 76),
 (1956, 74),
 (1955, 70),
 (1953, 63),
 (1954, 54),
 (2005, 46),
 (1951, 40),
 (1940, 36),
 (1952, 34),
 (1920, 33),
 (1950, 32),
 (1911, 19),
 (1942, 14),
 (1947, 14),
 (1946, 13),
 (1930, 13),
 (1923, 11),
 (1949, 11),
 (1941, 10),
 (1939, 9),
 (1948, 9),
 (1943, 8),
 (1945, 8),
 (1938, 7),
 (1901, 7),
 (1936, 7),
 (1929, 7),
 (1937, 5),
 (1932, 5),
 (1933, 4),
 (1944, 4),
 (1931, 3),
 (1935, 3),
 (2006, 3),
 (2020, 3),
 (1900, 3),
 (1927, 2),
 (1928, 2),
 (1926, 2),
 (1925, 2),
 (1924, 2),
 (1922, 2),
 (1921, 2),
 (1909, 2),
 (1902, 2),
 (2010, 2),
 (2011, 2),
 (1904, 1),
 (2012, 1),
 (1914, 1),
 (1917, 1),
 (1897, 1),
 (1806, 1),
 (1378, 1),
 (1919, 1),
 (1934, 1),
 (2008, 1),
 (1376, 1),
 (1910, 1),
 (2021, 1),
 (1908, 1),
 (1906, 1)]

### 7. Which publishers release the most books per year?

In [108]:
%%sql
-- query which publishers release the most books
select publisher, count(publisher)
from books
group by publisher
order by count(publisher) desc
limit 15;

 * postgresql://postgres:***@localhost/postgres
15 rows affected.


[('Harlequin', 7536),
 ('Silhouette', 4220),
 ('Pocket', 3905),
 ('Ballantine Books', 3783),
 ('Bantam Books', 3647),
 ('Scholastic', 3160),
 ('Simon &amp; Schuster', 2971),
 ('Penguin Books', 2844),
 ('Berkley Publishing Group', 2771),
 ('Warner Books', 2727),
 ('Penguin USA', 2717),
 ('Harpercollins', 2526),
 ('Fawcett Books', 2258),
 ('Signet Book', 2070),
 ('Random House Inc', 2045)]

There are some spelling differences that could affect this query as well, but once again, that data cleaning is outside the scope of this project.

### 8. Which publishers are rated the highest and lowest?

In [118]:
%%sql
-- select publisher with highest average rating and more than 50 reviews
select publisher, avg(rating) as avg_rating, count(rating) as num_of_ratings
from books b
inner join ratings r on b.isbn = r.isbn
group by publisher
having count(rating) > 50
order by avg(rating) desc
limit 1;

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[('Five Star (ME)', Decimal('7.8840579710144928'), 207)]

In [119]:
%%sql
-- select publisher with lowest average rating and more than 50 reviews
select publisher, avg(rating) as avg_rating, count(rating) as num_of_ratings
from books b
inner join ratings r on b.isbn = r.isbn
group by publisher
having count(rating) > 50
order by avg(rating)
limit 1;

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[('Meteor Publishing Corporation', Decimal('0.79166666666666666667'), 96)]

### 9. Which authors have the most books?

In [160]:
%%sql
select book_author, count(isbn)
from books
group by book_author
order by count(isbn) desc
limit 15;

 * postgresql://postgres:***@localhost/postgres
15 rows affected.


[('Agatha Christie', 632),
 ('William Shakespeare', 567),
 ('Stephen King', 524),
 ('Ann M. Martin', 423),
 ('Francine Pascal', 373),
 ('Carolyn Keene', 373),
 ('Isaac Asimov', 330),
 ('Nora Roberts', 315),
 ('Barbara Cartland', 307),
 ('Charles Dickens', 302),
 ('Not Applicable (Na )', 286),
 ('R. L. Stine', 282),
 ('Mark Twain', 231),
 ('Jane Austen', 223),
 ('Terry Pratchett', 220)]

### 10. Which authors have the most reviews? Is there any correlation to the authors with the most books?

We will take a round-about approach here so that we can create a view and show that functionality.

In [125]:
%%sql
-- query which authors have the most ratings
select b.book_author, count(r.rating) as num_of_ratings
from books b
inner join ratings r on b.isbn = r.isbn
group by book_author
order by count(r.rating) desc
limit 15;

 * postgresql://postgres:***@localhost/postgres
15 rows affected.


[('Stephen King', 10053),
 ('Nora Roberts', 8429),
 ('John Grisham', 6010),
 ('James Patterson', 5845),
 ('Mary Higgins Clark', 4777),
 ('Dean R. Koontz', 4313),
 ('Tom Clancy', 4036),
 ('Danielle Steel', 3726),
 ('Sue Grafton', 3457),
 ('Janet Evanovich', 3350),
 ('Anne Rice', 3030),
 ('Michael Crichton', 2921),
 ('J. K. Rowling', 2908),
 ('JOHN GRISHAM', 2808),
 ('V.C. Andrews', 2785)]

In [130]:
%%sql
-- create a view with the query information of authors and their rating counts
create view rating_counts as
select b.book_author, count(r.rating) as num_of_ratings
from books b
inner join ratings r on b.isbn = r.isbn
group by book_author;

 * postgresql://postgres:***@localhost/postgres
Done.


[]

In [164]:
%%sql
select b.book_author, b.num_of_books, r.num_of_ratings
from (select book_author, count(isbn) as num_of_books 
      from books
      group by book_author
     ) as b
inner join rating_counts r on b.book_author = r.book_author
order by num_of_books desc
limit 30;

 * postgresql://postgres:***@localhost/postgres
30 rows affected.


[('Agatha Christie', 632, 2344),
 ('William Shakespeare', 567, 1751),
 ('Stephen King', 524, 10053),
 ('Ann M. Martin', 423, 1938),
 ('Carolyn Keene', 373, 1076),
 ('Francine Pascal', 373, 1252),
 ('Isaac Asimov', 330, 975),
 ('Nora Roberts', 315, 8429),
 ('Barbara Cartland', 307, 341),
 ('Charles Dickens', 302, 1162),
 ('Not Applicable (Na )', 286, 445),
 ('R. L. Stine', 282, 2606),
 ('Mark Twain', 231, 891),
 ('Jane Austen', 223, 1557),
 ('Terry Pratchett', 220, 1572),
 ('Mary Higgins Clark', 218, 4777),
 ('Piers Anthony', 217, 1799),
 ('Marion Zimmer Bradley', 216, 1110),
 ('Janet Dailey', 214, 687),
 ('Franklin W. Dixon', 204, 471),
 ('Roald Dahl', 195, 1205),
 ('Dick Francis', 174, 1439),
 ('J. R. R. Tolkien', 174, 978),
 ('Tom Clancy', 174, 4036),
 ('Danielle Steel', 173, 3726),
 ('John Steinbeck', 172, 1174),
 ('Anne McCaffrey', 172, 1934),
 ('Nathaniel Hawthorne', 165, 437),
 ('Fern Michaels', 165, 1378),
 ('Enid Blyton', 164, 287)]

If we edit this query to show more entries, we will quickly see that there are some authors with more reviews and fewer books. While there is a general correlation (that should be expected) that higher publication count leads to more reviews, there are also other factors that make this far from perfect. Things like unclean data and popularity likely play a role here. We could also ask questions on genre demographics that would contribute to lesser correlation as well.

### Bonus: Let's use a window function to show average ratings per author against individual book data.

We will doctor the query a touch more to show some middle of the pack data rather than the outliers. 

In [187]:
%%sql
with windowed as(
    select b.book_author, b.book_title,
           round(avg(r.rating) over(partition by(b.book_author)), 2) as avg_author_rating
    from books b
    inner join ratings r on b.isbn = r.isbn
)

select * from windowed
where avg_author_rating < 8 and avg_author_rating > 7
limit 20;

 * postgresql://postgres:***@localhost/postgres
20 rows affected.


[('A. E. (Alfred Elton) Van Vogt', 'Le Silkie', Decimal('7.50')),
 ('A. E. (Alfred Elton) Van Vogt', 'Le Silkie', Decimal('7.50')),
 ('A.J. Chevalier', 'What If . . .', Decimal('7.50')),
 ('A.J. Chevalier', 'What If . . .', Decimal('7.50')),
 ('ADELINE YEN MAH', 'Chinese Cinderella : The true story of an unwanted daughter', Decimal('7.50')),
 ('ADELINE YEN MAH', 'Chinese Cinderella : The true story of an unwanted daughter', Decimal('7.50')),
 ('ALEXANDER HAMILTON', 'The Federalist : A Commentary on the Constitution of the United States (Modern Library Classics)', Decimal('7.50')),
 ('ALEXANDER HAMILTON', 'The Federalist : A Commentary on the Constitution of the United States (Modern Library Classics)', Decimal('7.50')),
 ('ALLEN KLEIN', 'Up Words for Down Days', Decimal('7.67')),
 ('ALLEN KLEIN', 'Up Words for Down Days', Decimal('7.67')),
 ('ALLEN KLEIN', 'Lift-Your-Spirits Quote Book', Decimal('7.67')),
 ('APRIL FROST', 'Beyond Obedience : Training with Awareness for You &amp; Your Dog', Decimal('7.50')),
 ('APRIL FROST', 'Beyond Obedience : Training with Awareness for You &amp; Your Dog', Decimal('7.50')),
 ('Adam Jamrozik', 'The Sociology of Social Problems : Theoretical Perspectives and Methods of Intervention', Decimal('7.50')),
 ('Adam Jamrozik', 'Class, Inequality and the State: Social Change, Social Policy and the New Middle Class', Decimal('7.50')),
 ('Adam Zickerman', 'Power of 10: The Once-a-Week, Slow Motion Fitness Revolution', Decimal('7.50')),
 ('Adam Zickerman', 'Power of 10: The Once-a-Week, Slow Motion Fitness Revolution', Decimal('7.50')),
 ('Akhil Sharma', 'An Obedient Father', Decimal('7.25')),
 ('Akhil Sharma', 'An Obedient Father', Decimal('7.25')),
 ('Akhil Sharma', 'An Obedient Father', Decimal('7.25'))]

This concludes this basic SQL exercise.

In [190]:
%%sql
select * from books
limit 5;

 * postgresql://postgres:***@localhost/postgres
5 rows affected.


[('0195153448', 'Classical Mythology', 'Mark P. O. Morford', 2002, 'Oxford University Press', 'http://images.amazon.com/images/P/0195153448.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0195153448.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0195153448.01.LZZZZZZZ.jpg'),
 ('0002005018', 'Clara Callan', 'Richard Bruce Wright', 2001, 'HarperFlamingo Canada', 'http://images.amazon.com/images/P/0002005018.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0002005018.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0002005018.01.LZZZZZZZ.jpg'),
 ('0060973129', 'Decision in Normandy', "Carlo D'Este", 1991, 'HarperPerennial', 'http://images.amazon.com/images/P/0060973129.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0060973129.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0060973129.01.LZZZZZZZ.jpg'),
 ('0374157065', 'Flu: The Story of the Great Influenza Pandemic of 1918 and the Search for the Virus That Caused It', 'Gina Bari Kolata', 1999, 'Farrar Straus Giroux', 'http://images.amazon.com/images/P/0374157065.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0374157065.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0374157065.01.LZZZZZZZ.jpg'),
 ('0393045218', 'The Mummies of Urumchi', 'E. J. W. Barber', 1999, 'W. W. Norton &amp; Company', 'http://images.amazon.com/images/P/0393045218.01.THUMBZZZ.jpg', 'http://images.amazon.com/images/P/0393045218.01.MZZZZZZZ.jpg', 'http://images.amazon.com/images/P/0393045218.01.LZZZZZZZ.jpg')]